drivers/clock_control/clock_stm32_ll_u3.c

```diff
@@ -0,0 +1,603 @@
+#include <zephyr/arch/cpu.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/drivers/clock_control/stm32_clock_control.h>
+
+
+/* Macros to fill up prescaler values */
+#define z_ahb_prescaler(v) LL_RCC_HCLK_SYSCLK_DIV_ ## v
+#define ahb_prescaler(v) z_ahb_prescaler(v)
+
+#define z_apb1_prescaler(v) LL_RCC_APB1_HCLK_DIV_ ## v
+#define apb1_prescaler(v) z_apb1_prescaler(v)
+
+#define z_apb2_prescaler(v) LL_RCC_APB2_HCLK_DIV_ ## v
+#define apb2_prescaler(v) z_apb2_prescaler(v)
+
+#define z_apb3_prescaler(v) LL_RCC_APB3_HCLK_DIV_ ## v
+#define apb3_prescaler(v) z_apb3_prescaler(v)
+
+static uint32_t get_bus_clock(uint32_t clock, uint32_t prescaler)
+{
+	return clock / prescaler;
+}
+
+static uint32_t get_msis_frequency(void)
+{
+	uint32_t frequency = MSIRC1_VALUE;
+
+	if (LL_RCC_MSIS_GetClockSource() == LL_RCC_MSIS_CLOCK_SOURCE_RC0)
+	{
+	frequency = MSIRC0_VALUE;
+	}
+	switch (LL_RCC_MSIS_GetClockDivision())
+	{
+	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_2:
+	frequency /= 2U;
+	break;
+	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_4:
+	frequency /= 4U;
+	break;
+	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_8:
+	frequency /= 8U;
+	break;
+	case LL_RCC_MSIS_CLOCK_SOURCE_RC_DIV_1:
+	break;
+	default:
+	return 0;
+	}
+
+	return frequency;
+}
+
+static uint32_t get_startup_frequency(void)
+{
+	switch (LL_RCC_GetSysClkSource()) {
+	case LL_RCC_SYS_CLKSOURCE_STATUS_MSIS:
+		return get_msis_frequency();
+	case LL_RCC_SYS_CLKSOURCE_STATUS_HSI16:
+		return STM32_HSI_FREQ;
+	case LL_RCC_SYS_CLKSOURCE_STATUS_HSE:
+		return STM32_HSE_FREQ;
+	default:
+		__ASSERT(0, "Unexpected startup freq");
+		return 0;
+	}
+}
+
+static uint32_t get_sysclk_frequency(void)
+{
+#if defined(STM32_SYSCLK_SRC_MSIS)
+	return get_msis_frequency();
+#elif defined(STM32_SYSCLK_SRC_HSE)
+	return STM32_HSE_FREQ;
+#elif defined(STM32_SYSCLK_SRC_HSI)
+	return STM32_HSI_FREQ;
+#else
+	__ASSERT(0, "No SYSCLK Source configured");
+	return 0;
+#endif
+
+}
+
+/** @brief Verifies clock is part of active clock configuration */
+int enabled_clock(uint32_t src_clk)
+{
+	if ((src_clk == STM32_SRC_SYSCLK) ||
+	    (src_clk == STM32_SRC_HCLK) ||
+	    (src_clk == STM32_SRC_PCLK1) ||
+	    (src_clk == STM32_SRC_PCLK2) ||
+	    (src_clk == STM32_SRC_PCLK3) ||
+	    ((src_clk == STM32_SRC_HSE) && IS_ENABLED(STM32_HSE_ENABLED)) ||
+	    ((src_clk == STM32_SRC_HSI16) && IS_ENABLED(STM32_HSI_ENABLED)) ||
+	    ((src_clk == STM32_SRC_HSI48) && IS_ENABLED(STM32_HSI48_ENABLED)) ||
+	    ((src_clk == STM32_SRC_LSE) && IS_ENABLED(STM32_LSE_ENABLED)) ||
+	    ((src_clk == STM32_SRC_LSI) && IS_ENABLED(STM32_LSI_ENABLED)) ||
+	    ((src_clk == STM32_SRC_MSIS) && IS_ENABLED(STM32_MSIS_ENABLED)) ||
+	    ((src_clk == STM32_SRC_MSIK) && IS_ENABLED(STM32_MSIK_ENABLED))
+	) {
+		return 0;
+	}
+
+	return -ENOTSUP;
+}
+
+static int stm32_clock_control_on(const struct device *dev, clock_control_subsys_t sub_system)
+{
+	struct stm32_pclken *pclken = (struct stm32_pclken *)(sub_system);
```

Comment:
Could remove the parenthesis around `sub_system`?
Ditto in the other local functions below.